### PR TITLE
HSEARCH-4112 Hibernate Search + Spring-JTA fails because of closed entity manager

### PR DESCRIPTION
**Pull request by yrodiere**
https://hibernate.atlassian.net/browse/HSEARCH-4112

### CODE REVIEW

File: mapper/orm/src/main/java/org/hibernate/search/mapper/orm/session/impl/HibernateOrmSearchSession.java [hunk 1]
```diff
@@ -249,27 +245,23 @@ public PojoRuntimeIntrospector runtimeIntrospector() {
...
-
-		TransientReference<Map<Transaction, PojoIndexingPlan<EntityReference>>> reference =
-				(TransientReference<Map<Transaction, PojoIndexingPlan<EntityReference>>>) sessionImplementor.getProperties()
-				.get( INDEXING_PLAN_PER_TRANSACTION_MAP_KEY );
-		Map<Transaction, PojoIndexingPlan<EntityReference>> planPerTransaction = reference == null ? null : reference.get();
-		if ( planPerTransaction == null ) {
-			planPerTransaction = new HashMap<>();
-			reference = new TransientReference<>( planPerTransaction );
-			sessionImplementor.setProperty( INDEXING_PLAN_PER_TRANSACTION_MAP_KEY, reference );
+		HibernateOrmSearchSessionHolder properties =
```
Comment (fax4ever) [inline mapper/orm/src/main/java/org/hibernate/search/mapper/orm/session/impl/HibernateOrmSearchSession.java:249]: This is not critical, but I think that `properties` is not a very clear name for the holder instance.

File: mapper/orm/src/main/java/org/hibernate/search/mapper/orm/session/impl/HibernateOrmSearchSessionHolder.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+		if ( holder != null ) {
+			return holder;
+		}
+		boolean closedSessionAndTransaction = session.isClosed() && session.isTransactionInProgress();
+		if ( closedSessionAndTransaction ) {
+			// This can happen when using JTA.
+			// Only do this as a fallback: if somehow the session holder was added to the session properties
+			// before the session was closed, we definitely want to use it and avoid the static map.
+			holder = holderPerClosedSessionTransaction.get( session.accessTransaction() );
+		}
```
Comment (fax4ever) [inline mapper/orm/src/main/java/org/hibernate/search/mapper/orm/session/impl/HibernateOrmSearchSessionHolder.java:49]: This is not critical either, the code looks good in any case, but maybe we can introduce 2 sub classes: one for JTA and the other for the non-JTA case.

**Review by fax4ever — APPROVED** [review]
Really a nice job! +1 to merge it
Just very optional comments.

File: mapper/orm/src/main/java/org/hibernate/search/mapper/orm/session/impl/HibernateOrmSearchSessionHolder.java (hunk 2 above)
Comment (yrodiere) [reply mapper/orm/src/main/java/org/hibernate/search/mapper/orm/session/impl/HibernateOrmSearchSessionHolder.java:49]: This is a static method, no polymorphism here.

Also, we will not always end up in this branch, even with JTA: if a `SearchSession` was retrieved before the commit, we'll proceed with the "normal" behavior (which is better).

File: mapper/orm/src/main/java/org/hibernate/search/mapper/orm/session/impl/HibernateOrmSearchSession.java (hunk 1 above)
Comment (yrodiere) [reply mapper/orm/src/main/java/org/hibernate/search/mapper/orm/session/impl/HibernateOrmSearchSession.java:249]: Right, will do

**Comment by yrodiere** [comment]
Merged, thanks

**Pull request merged**